django/db/models/window.py

```diff
@@ -0,0 +1,203 @@
+from django.db.models import Expression, F
+from django.db.models.expressions import ExpressionList, OrderBy
+
+__all__ = ['RowRange', 'ValueRange', 'Window']
+
+
+class Window(Expression):
+    template = '%(expression)s OVER (%(window)s)'
+    filterable = False
+
+    def __init__(self, expression, partition_by=None, order_by=None, frame=None, output_field=None):
+        self.frame = frame
+        self.partition_by = partition_by
+        self.order_by = order_by
+
+        if self.partition_by is not None:
+            if isinstance(self.partition_by, (tuple, list)):
+                self.partition_by = ExpressionList(*self.partition_by)
+            else:
+                self.partition_by = ExpressionList(self.partition_by)
+
+        if isinstance(self.order_by, str):
+            if order_by.startswith('-'):
+                self.order_by = OrderBy(F(self.order_by[1:]), descending=True)
+            else:
+                self.order_by = OrderBy(F(self.order_by))
+
+        super().__init__(output_field=output_field)
+        self.source_expression = self._parse_expressions(expression)[0]
+        if not getattr(self.source_expression, 'contains_aggregate', False):
+            raise FieldError('Window function expressions must be aggregate functions')
+
+    def contains_aggregate(self):
+        # although the expression contains an aggregate function, the ORM
+        # handles such expressions in ways that may not always be ideal for an
+        # OVER-clause
+        return False
+
+    def _resolve_output_field(self):
+        if self._output_field is None:
+            self._output_field = self.source_expression.output_field
+
+    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
+        c = self.copy()
+        c.source_expression = c.source_expression.resolve_expression(query, allow_joins, reuse, summarize, for_save)
+        c.is_summary = summarize
+        c.for_save = for_save
+
+        if c.partition_by:
+            c.partition_by = c.partition_by.resolve_expression(query, allow_joins, reuse, summarize, for_save)
+        if c.order_by:
+            c.order_by = c.order_by.resolve_expression(query, allow_joins, reuse, summarize, for_save)
+        if c.frame:
+            c.frame = c.frame.resolve_expression(query, allow_joins, reuse, for_save)
+        return c
+
+    def as_sql(self, compiler, connection, function=None, template=None):
+        connection.ops.check_expression_support(self)
+        expr_sql, params = compiler.compile(self.source_expression)
+        window_sql, window_params = [], []
+
+        if self.partition_by is not None:
+            sql_expr, sql_params = self.partition_by.as_sql(
+                compiler=compiler, connection=connection, template='PARTITION BY %(expressions)s')
+            window_sql.extend(sql_expr)
+            window_params.extend(sql_params)
+
+        if self.order_by is not None:
```

Comment:
Reminder to self: check if sqli is possible here. It doesn't *look* like it is, but I want to double check.